### DISCUSSION TITLE
Emit data while AsyncLoading

### DISCUSSION THREAD

**Discussion by csbence10** [rrousselGit/riverpod#2588] [Q&A] [answered]
Hey!

I want to get filtered data from a remote server. I have a `Notifier` managing the filtering, and I also have an `FutureProvider` fetching the data based on the filter. And - **the important part** - I want to show the data (if any) while the provider is refreshing due to filter changes.

So it's something like that:
```
@riverpod
Future<MyData> remoteData(RemoteDataRef ref) {
  final myFilter = ref.watch(myFilterProvider);
  return ref.watch(apiClient).getData(filter: myFilter);
}
```

So far, so good. Now I want to filter the returned data further so I define more providers like:
 ```
@riverpod
AsyncValue<MyFilteredData> filteredData(FilteredDataRef ref) {
  final data = ref.watch(remoteDataProvider);
  return data.whenData((data) => data.someProperty.map(someMapper));
}
 ```

I use `AsyncValue.when` on the UI to switch the states. When I change the state of `myFilterProvider` the loading case is returned and this is what I don't want. I tried to set `skipLoadingOnReload: true` but that didn't help.

Now my only solution is this one, which I don't really like:

```
@riverpod
AsyncValue<MyFilteredData> filteredData(FilteredDataRef ref) {
 return ref.watch(remoteDataProvider).map(
        data: (data) => AsyncData(data.someProperty.map(someMapper)),
        error: (error) => AsyncError(error, error.stackTrace),
        loading: (loading) {
          if (loading.isReloading && loading.value != null) {
            return AsyncData(data.someProperty.map(someMapper));
          }
          return const AsyncLoading();
        },
      );
}
```

Anyone has any idea how to solve this a shorter, less ugly way? Thanks in advance

**Comment by AhmedLSayed9** [comment, marked as answer]
> @riverpod
> AsyncValue<MyFilteredData> filteredData(FilteredDataRef ref) {
>  final data = ref.watch(remoteDataProvider);
>  return data.whenData((data) => data.someProperty.map(someMapper));
> }

`whenData` only handle (mapping) the `data` case and transit the error/loading cases as-is, but they lose the prev data/error (They return as raw AsyncLoading/AsyncError). That's why `skipLoadingOnReload: true` doesn't work for you on the UI.

I think they should preserve prev data/error anyway as `whenData` is commonly used for transiting from some AsyncValue to another. What do you think @rrousselGit ?

Anyway, you can solve this by either skip loading at your `filteredDataProvider` instead of the UI:
```dart
@riverpod
AsyncValue<MyFilteredData> filteredData(FilteredDataRef ref) {
  return ref.watch(remoteDataProvider).when(
    skipLoadingOnReload: true,
    data: (d) => AsyncData(d.someProperty.map(someMapper)),
    error: (err, st) => AsyncError(err, st),
    loading: () => const AsyncLoading(),
  );
}
```
or, stop manual transition to a new `AsyncValue` and use `await + .future`, then `skipLoadingOnReload` will work fine on the UI:
```dart
@riverpod
MyFilteredData filteredData(FilteredDataRef ref) async {
  final data = await ref.watch(remoteDataProvider.future);
  return data.someProperty.map(someMapper);
}
```
I'd go with second option.

**Reply by rrousselGit** [reply]
The .future variant is what's expected.   
You're not really meant to return an AsyncValue in a provider. If you do that, you'll have to handle previous values on your own.

**Reply by csbence10** [reply]
Thanks for the quick answer. Works like a charm 🙏
